rust/noosphere-storage/src/implementation/rocks_db.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{storage::Storage, store::Store, SPHERE_DB_STORE_NAMES};
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use rocksdb::{ColumnFamilyDescriptor, DBWithThreadMode, Options};
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+#[cfg(not(feature = "rocksdb-multi-thread"))]
+type DbInner = DBWithThreadMode<rocksdb::SingleThreaded>;
+#[cfg(not(feature = "rocksdb-multi-thread"))]
+type ColumnType<'a> = &'a rocksdb::ColumnFamily;
+#[cfg(feature = "rocksdb-multi-thread")]
+type DbInner = DBWithThreadMode<rocksdb::MultiThreaded>;
+#[cfg(feature = "rocksdb-multi-thread")]
+type ColumnType<'a> = Arc<rocksdb::BoundColumnFamily<'a>>;
+
+/// A RocksDB implementation of [Storage].
+///
+/// Caveats:
+/// * Values are limited to 4GB(?) [https://github.com/facebook/rocksdb/wiki/Basic-Operations#reads]
+/// * Potential workarounds of cf_handles not being [Sync] [https://github.com/rust-rocksdb/rust-rocksdb/issues/407]
+///   Could potentially use `noosphere_common::Unshared`?
+#[derive(Clone, Debug)]
+pub struct RocksDbStorage {
+    db: Arc<DbInner>,
+    #[allow(unused)]
+    path: PathBuf,
+}
+
+impl RocksDbStorage {
+    pub fn new(path: PathBuf) -> Result<Self> {
```

Comment:
What happens if the storage is initialized multiple times over the same location? Idempotent/safe to do?

Reply:
Same restrictions as sled e.g. only one `new()` per `path` allowed at a time, e.g. this fails:

```rs
    async fn it_can_open_concurrent_dbs() -> Result<()> {
        let tempdir = tempfile::TempDir::new()?;
        let storage_path = tempdir.path().to_owned();

        let storage_1 = RocksDbStorage::new(storage_path.clone())?;
        let mut store_1 = storage_1.get_key_value_store("links").await?;
        let storage_2 = RocksDbStorage::new(storage_path.clone())?;
        let mut store_2 = storage_2.get_key_value_store("links").await?;

        let bytes_1 = vec![0, 1, 2, 3];
        let bytes_2 = vec![4, 5, 6, 7];
        store_1.write(b"from 1", &bytes_1).await?;
        store_2.write(b"from 2", &bytes_2).await?;
        assert_eq!(store_1.read(b"from 2").await?.unwrap(), bytes_2);
        assert_eq!(store_2.read(b"from 1").await?.unwrap(), bytes_1);
        Ok(())
    }
```